transport-native-kqueue/src/main/java/io/netty/channel/kqueue/KQueueSocketChannel.java

```diff
@@ -63,6 +67,32 @@ public ServerSocketChannel parent() {
         return (ServerSocketChannel) super.parent();
     }
 
+    @Override
+    protected boolean doConnect0(SocketAddress remoteAddress, SocketAddress localAddress) throws Exception {
+        if (config.isTcpFastOpenConnect()) {
+            ChannelOutboundBuffer outbound = unsafe().outboundBuffer();
+            outbound.addFlush();
+            Object curr;
+            if ((curr = outbound.current()) instanceof ByteBuf) {
+                ByteBuf initialData = (ByteBuf) curr;
+                if (initialData.isReadable()) {
+                    IovArray iov = new IovArray(config.getAllocator().directBuffer());
+                    try {
+                        iov.add(initialData, initialData.readerIndex(), initialData.readableBytes());
+                        int bytesSent = socket.connectx(
+                                (InetSocketAddress) localAddress, (InetSocketAddress) remoteAddress, iov, true);
+                        writeFilter(true);
+                        outbound.removeBytes(bytesSent);
+                        return false; // 'false' because we assume connecting to be in-progress.
```

Comment:
Aren't we losing information here because `connectx` could have already succeeded? The downside is that we will do additional work in [`AbstractKqueueUnsafe#connect()`](https://github.com/netty/netty/blob/4.1/transport-native-kqueue/src/main/java/io/netty/channel/kqueue/AbstractKQueueChannel.java#L548-L552) if we return `false` here when the connection is already established.

Reply:
I can try to encode it in a long.